### PR TITLE
feat: FromPretrained to load tokenizer directly from HF

### PR DESCRIPTION
**Pull request by berkayersoyy**
| Benchmark            | benchmarks/eafe2ce1db50dcd78b3b7157bf2c669109f4e322.txt | benchmarks/867b67eeade51e570d5150cebddab679b855e885.txt | Change               |
|----------------------|--------------------------------------------------------|--------------------------------------------------------|----------------------|
| **sec/op**           |                                                        |                                                        |                      |
| EncodeNTimes-10      | 12.75µ ± 2%                                            | 12.60µ ± 1%                                            | -1.18% (p=0.015 n=6) |
| EncodeNChars-10      | 0.3629n ± 19%                                          | 0.3589n ± 8%                                           | ~ (p=1.000 n=6)      |
| DecodeNTimes-10      | 4.433µ ± 1%                                            | 4.481µ ± 1%                                            | +1.08% (p=0.011 n=6) |
| DecodeNTokens-10     | 660.5n ± 2%                                            | 662.0n ± 5%                                            | ~ (p=0.485 n=6)      |
| Geomean              | 341.2n                                                 | 340.3n                                                 | -0.25%               |

| Benchmark            | benchmarks/eafe2ce1db50dcd78b3b7157bf2c669109f4e322.txt | benchmarks/867b67eeade51e570d5150cebddab679b855e885.txt | Change               |
|----------------------|--------------------------------------------------------|--------------------------------------------------------|----------------------|
| **B/op**             |                                                        |                                                        |                      |
| EncodeNTimes-10      | 232.0 ± 0%                                             | 232.0 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| EncodeNChars-10      | 0.000 ± 0%                                             | 0.000 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| DecodeNTimes-10      | 96.00 ± 0%                                             | 96.00 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| DecodeNTokens-10     | 7.000 ± 0%                                             | 7.000 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| Geomean              |                                                        |                                                        | +0.00%²              |

| Benchmark            | benchmarks/eafe2ce1db50dcd78b3b7157bf2c669109f4e322.txt | benchmarks/867b67eeade51e570d5150cebddab679b855e885.txt | Change               |
|----------------------|--------------------------------------------------------|--------------------------------------------------------|----------------------|
| **allocs/op**        |                                                        |                                                        |                      |
| EncodeNTimes-10      | 12.00 ± 0%                                             | 12.00 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| EncodeNChars-10      | 0.000 ± 0%                                             | 0.000 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| DecodeNTimes-10      | 3.000 ± 0%                                             | 3.000 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| DecodeNTokens-10     | 0.000 ± 0%                                             | 0.000 ± 0%                                             | ~ (p=1.000 n=6)¹     |
| Geomean              |                                                        |                                                        | +0.00%²              |

¹ All samples are equal  
² Summaries must be >0 to compute geomean

### CODE REVIEW

**Comment by berkayersoyy** [comment]
I added a few things so that the tokenizer can be loaded directly from huggingface, if you could review it that would be great @daulet 🙏

File: tokenizer.go [hunk 1]
```diff
@@ -67,6 +88,155 @@ func FromFile(path string) (*Tokenizer, error) {
...
+		// Create a temporary directory
+		tmpDir, err := os.MkdirTemp("", "huggingface-tokenizer-*")
+		if err != nil {
+			return nil, fmt.Errorf("error creating temporary directory: %w", err)
+		}
+		downloadDir = tmpDir
+	}
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(tokenizerFiles))
```
Comment (daulet) [inline tokenizer.go:124]: let's use unbuffered channel: https://github.com/uber-go/guide/blob/master/style.md#channel-size-is-one-or-none

using buffered channels tends to hide concurrency issues if present

File: tokenizer.go [hunk 2]
```diff
@@ -67,6 +88,155 @@ func FromFile(path string) (*Tokenizer, error) {
...
+			fileURL := fmt.Sprintf("%s/%s", modelURL, fn)
+			destPath := filepath.Join(downloadDir, fn)
+			err := downloadFile(fileURL, destPath, authToken)
+			if err != nil {
+				if mandatory {
+					// If the file is mandatory, report an error
+					errChan <- fmt.Errorf("failed to download mandatory file %s: %w", fn, err)
+				} else {
+					// Optional files: log warning and continue
+					logMutex.Lock()
```
Comment (daulet) [inline tokenizer.go:143]: i dont believe you need this mutex at all

> fmt::print provides the same thread-safety guarantees as the underlying system API which is fwrite (or WriteConsoleW when writing to the console on Windows) for FILE* overloads. These APIs are normally thread-safe

File: tokenizer.go [hunk 3]
```diff
@@ -67,6 +88,155 @@ func FromFile(path string) (*Tokenizer, error) {
...
+		return nil, fmt.Errorf("mandatory file tokenizer.json does not exist in %s", downloadDir)
+	}
+
+	// Initialize the tokenizer using the downloaded tokenizer.json
+	tokenizer, err := FromFile(tokenizerPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return tokenizer, nil
```
Comment (daulet) [inline tokenizer.go:176]:
```suggestion
	return FromFile(tokenizerPath)
```

File: tokenizer.go [hunk 4]
```diff
@@ -67,6 +88,155 @@ func FromFile(path string) (*Tokenizer, error) {
...
+					fmt.Printf("Warning: failed to download optional file %s: %v\n", fn, err)
+					logMutex.Unlock()
+				}
+			}
+		}(filename, isMandatory)
+	}
+
+	// Wait for all downloads to complete
+	wg.Wait()
+	close(errChan)
```
Comment (daulet) [inline tokenizer.go:153]: nit: name this `errCh`

File: tokenizer.go [hunk 5]
```diff
@@ -67,6 +88,155 @@ func FromFile(path string) (*Tokenizer, error) {
...
+
+	// Check for errors during downloads
+	for downloadErr := range errChan {
+		if downloadErr != nil {
```
Comment (daulet) [inline tokenizer.go:157]: looks like all errors sent to the channel are non-nil so check is redundant. Just looking out for minimal indentation

**Review by daulet — COMMENTED** [review]
Thank you for your contribution! In principle looks good, let's just iron out concurrency stuff.

**Comment by berkayersoyy** [comment]
I made the changes you mentioned, can you review them again? 🙏

File: tokenizer.go [hunk 6]
```diff
@@ -67,6 +88,142 @@ func FromFile(path string) (*Tokenizer, error) {
 	return &Tokenizer{tokenizer: tokenizer}, nil
 }
 
+// FromPretrained downloads necessary files and initializes the tokenizer.
+// Parameters:
+//   - modelID: The Hugging Face model identifier (e.g., "bert-base-uncased").
+//   - destination: Optional. If provided and not nil, files will be downloaded to this folder.
+//     If nil, a temporary directory will be used.
+//   - authToken: Optional. If provided and not nil, it will be used to authenticate requests.
+func FromPretrained(modelID string, destination, authToken *string) (*Tokenizer, error) {
```
Comment (daulet) [inline tokenizer.go:97]: let's add a test to exercise default and optional params, e.g. with and without optional auth token

**Comment by daulet** [comment]
@berkayersoyy I've made few changes instead of PR comments, lets just add tests and we are good to go

**Comment by berkayersoyy** [comment]
looks good 🙏 , I added the tests @daulet

File: tokenizer_test.go [hunk 7]
```diff
@@ -2,18 +2,39 @@ package tokenizers_test
 
 import (
 	_ "embed"
+	"github.com/daulet/tokenizers"
```
Comment (daulet) [inline tokenizer_test.go:5]: this shouldn't be necessary, you are in the same package

File: tokenizer_test.go [hunk 8]
```diff
@@ -2,18 +2,39 @@ package tokenizers_test
...
 	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"strings"
 	"testing"
 
-	"github.com/daulet/tokenizers"
```
Comment (daulet) [inline tokenizer_test.go:8]: keep it in this order

File: tokenizer_test.go [hunk 9]
```diff
@@ -472,3 +493,250 @@ func BenchmarkDecodeNTokens(b *testing.B) {
...
+			"private/model",
+			tokenizers.WithAuthToken("test-token"),
+			tokenizers.WithCacheDir(t.TempDir()),
+		)
+
+		if err != nil {
+			t.Errorf("unexpected error with valid auth token: %v", err)
+			return
+		}
+		if tokenizer != nil {
```
Comment (daulet) [inline tokenizer_test.go:727]: redundant check

File: tokenizer_test.go [hunk 10]
```diff
@@ -472,3 +493,250 @@ func BenchmarkDecodeNTokens(b *testing.B) {
...
+		}
+
+		tokenizer, err = tokenizers.FromPretrained(
+			"private/model",
+			tokenizers.WithAuthToken("invalid-token"),
+			tokenizers.WithCacheDir(t.TempDir()),
+		)
+
+		if err == nil {
+			t.Error("expected error with invalid auth token, got nil")
```
Comment (daulet) [inline tokenizer_test.go:738]: rewrite these in table test format, where expected error, token are parameters, but the test body is the same

Comment (daulet) [reply tokenizer_test.go:738]: https://go.dev/wiki/TableDrivenTests

File: tokenizer_test.go [hunk 11]
```diff
@@ -472,3 +493,250 @@ func BenchmarkDecodeNTokens(b *testing.B) {
...
+
+			if tt.wantError {
+				if err == nil {
+					t.Error("expected error, got nil")
+				}
+				return
+			}
+
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
```
Comment (daulet) [inline tokenizer_test.go:659]: If you want Error + return, then probably just use `Fatalf`

**Review by daulet — CHANGES_REQUESTED** [review]
left comments

**Comment by berkayersoyy** [comment]
updated the tests 🙏 @daulet

**Review by daulet — APPROVED** [review]

**Comment by daulet** [comment]
Thank you @berkayersoyy for your contribution!

**Pull request merged**